### PR TITLE
rebase sgx mTLS to intel-22.06 release

### PR DESCRIPTION
**Pull request by Chen-Xintong**
**Please provide a description of this PR:**

Rebase internal sgx related code into 1.15 release.

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: manifests/charts/gateways/istio-egress/templates/deployment.yaml
Comment (ipuustin) [inline manifests/charts/gateways/istio-egress/templates/deployment.yaml:64]: Do I read this correctly that you need root access? Accessing devices from non-root containers should be have been fixed in both containerd and cri-o for a while now.

File: pilot/docker/Dockerfile.proxyv2
Comment (ipuustin) [inline pilot/docker/Dockerfile.proxyv2:10]: I think the latest release is 2.17.100.3.

File: pilot/docker/Dockerfile.proxyv2
Comment (ipuustin) [inline pilot/docker/Dockerfile.proxyv2:12]: 1.14.100.3 here.

File: pilot/pkg/xds/adstest.go
Comment (ipuustin) [inline pilot/pkg/xds/adstest.go:24]: Extra empty line.

**Review by ipuustin — CHANGES_REQUESTED** [review]

File: Makefile.core.mk
Comment (Chen-Xintong) [reply Makefile.core.mk]: Removed

File: manifests/charts/gateways/istio-egress/templates/deployment.yaml
Comment (Chen-Xintong) [reply manifests/charts/gateways/istio-egress/templates/deployment.yaml]: Yes, we shouldn't need root access, I removed root in ingress/egress and injection template deployment files, thanks for the correction.

File: pilot/docker/Dockerfile.proxyv2
Comment (Chen-Xintong) [reply pilot/docker/Dockerfile.proxyv2]: Okay, Fixed

File: pilot/docker/Dockerfile.proxyv2
Comment (Chen-Xintong) [reply pilot/docker/Dockerfile.proxyv2]: fixed

File: pilot/pkg/xds/adstest.go
Comment (Chen-Xintong) [reply pilot/pkg/xds/adstest.go]: fixed

File: pilot/docker/Dockerfile.proxyv2
Comment (ipuustin) [inline pilot/docker/Dockerfile.proxyv2:87]: Are you signing the enclave with the CTK test key? You need to sign using a real private key if someone needs to attest the quote (since the quote contains the MRSIGNER value). Is that the case since you build with `--enable-dcap`?

**Review by ipuustin — CHANGES_REQUESTED** [review]
Some comments but starting to look good. Can you rebase this after https://github.com/intel/istio/pull/4 gets merged so that we would see if CI can build this.

File: Makefile.core.mk
Comment (avalluri) [inline Makefile.core.mk:141]: Sorry if i miss something, are these makefile changes related to the SGX work? Do we really need them?

File: pilot/docker/Dockerfile.proxyv2
Comment (avalluri) [inline pilot/docker/Dockerfile.proxyv2:138]: SGX SDK and DCAP versions are different from the ones used in the build [above](https://github.com/intel/istio/pull/3/files#diff-3696302fd9b6db61d2c5b15bc3d35575ee7e35cf758aab7ef8a9e7e5b140e2beR10-R12) layer.

File: tools/docker.yaml
Comment (avalluri) [inline tools/docker.yaml:34]: I guess this line should not be removed. We should consider the RELEASE_MODE

File: security/pkg/testing/sdsc/sdsclient.go
Comment (avalluri) [inline security/pkg/testing/sdsc/sdsclient.go:27]: This new line could be removed.

File: manifests/charts/istio-control/istio-discovery/files/injection-template.yaml
Comment (Chen-Xintong) [reply manifests/charts/istio-control/istio-discovery/files/injection-template.yaml:370]: Yes, privileged can also be removed.

File: pilot/cmd/pilot-agent/options/security.go
Comment (Chen-Xintong) [reply pilot/cmd/pilot-agent/options/security.go:82]: If so then the certificate validator will not be enabled, and the `ValidatorConfig `will not passed to envoy.

File: pilot/docker/Dockerfile.proxyv2
Comment (Chen-Xintong) [reply pilot/docker/Dockerfile.proxyv2:87]: > Are you signing the enclave with the CTK test key? You need to sign using a real private key if someone needs to attest the quote (since the quote contains the MRSIGNER value). Is that the case since you build with `--enable-dcap`?

We are doing this temporarily. We will add a flag such as `--ctk-privkey="private key path"` in the build phase later, for users to choose to use their own key or the default key of CTK

File: manifests/charts/gateways/istio-egress/templates/deployment.yaml
Comment (Chen-Xintong) [reply manifests/charts/gateways/istio-egress/templates/deployment.yaml:162]: Am I missing something and why we don't need the sgx resource when the gateway has sgx enabled?

File: Makefile.core.mk
Comment (Chen-Xintong) [reply Makefile.core.mk:141]: Ahh, this was automatically migrated during cherry pick. Istio has modified the way docker builds before, so I think we don't need these anymore, thanks for correction.

**Comment by Chen-Xintong** [comment]
Hi @ipuustin, this PR has been modified and rebased according to your and @avalluri 's comments. Please merge this [PR](https://github.com/intel/envoy-go-control-plane/pull/1) of the generated sds proto first, and then wait until #4  merged, and then merge this.

File: manifests/charts/gateways/istio-egress/templates/deployment.yaml
Comment (avalluri) [reply manifests/charts/gateways/istio-egress/templates/deployment.yaml:162]: I meant to say was we are requesting only the `sgx.intel.com/epc` resource but `sgx.intel.com/enclave` resource is also required.

Comment (Chen-Xintong) [reply manifests/charts/gateways/istio-egress/templates/deployment.yaml:162]: As I known, once `sgx.intel.com/epc` is allocated, the `sgx.intel.com/enclave` resource will be allocated automatically. So we don't need to set the enclave resource separately.

File: security/pkg/pki/util/generate_cert.go
Comment (avalluri) [inline security/pkg/pki/util/generate_cert.go:439]: This new function is not used, could be removed?

File: manifests/charts/gateways/istio-egress/templates/deployment.yaml
Comment (avalluri) [reply manifests/charts/gateways/istio-egress/templates/deployment.yaml:162]: > As I known, once `sgx.intel.com/epc` is allocated, the `sgx.intel.com/enclave` resource will be allocated automatically. So we don't need to set the enclave resource separately.

Okay, thanks sharing this.

**Comment by ipuustin** [comment]
https://github.com/intel/envoy-go-control-plane/pull/1 is merged now, can you update `go.mod` here?

File: security/pkg/pki/util/generate_cert.go
Comment (Chen-Xintong) [reply security/pkg/pki/util/generate_cert.go]: Thanks, removed

**Comment by Chen-Xintong** [comment]
> [intel/envoy-go-control-plane#1](https://github.com/intel/envoy-go-control-plane/pull/1) is merged now, can you update `go.mod` here?

Updated.

**Comment by Chen-Xintong** [comment]
Hi @ipuustin, this PR has been updated and can successfully build binaries/containers on my host, but in the CI system of this repo, it seems that the `intel/envoy-go-control-plane` lib is not updated to this repo:  
`security/pkg/nodeagent/sds/sdsservice.go:323:24: undefined: sgxtls`
`security/pkg/nodeagent/sds/sdsservice.go:424:14: undefined: sgx`

Please check if the CI is updated, thanks.

**Comment by ZiyangXiao** [comment]
Seems that new proxyv2.dockerfile requires a proxy to work, @rmstoi @ipuustin How do we handle this in the internal repo？

**Comment by Chen-Xintong** [comment]
Thanks @ZiyangXiao , this CI check is now successful, @ipuustin , can you help merge this PR? thanks.

**Review by ipuustin — APPROVED** [review]

**Pull request merged**
